### PR TITLE
Filter patients whose second dose is overdue

### LINKED ISSUE #176: Filter patients whose second dose is overdue
Ability to use a toggle filter to query and fetch all patients who have missed their second shot and are overdue for the second dosage of the vaccine. 
Screenshot attached. This should only return results of all patient who have an  overdue status without the need to search 

Screen Shot 2021-05-28 at 12 54 21 PM

Screen Shot 2021-05-28 at 12 56 49 PM

**Some assumptions** 
1. Create a new patient on the registration page
2. Go to the profile page to record the first dose of the vaccine 
3. On record first dose, use current date as a start date when you received the 1st shot 
4. We shall use a 3-week window as the due date of the 2nd shot
5. On load patient list, if within the range before due date, show the user when you are due for the next shot, else show overdue status
6. Inpatient profile show when you were supposed to have received the 2nd shot - see screenshot. Label marked in red if overdue

Screen Shot 2021-05-28 at 1 08 16 PM

### PR DESCRIPTION
**Pull request by syedowaisali90**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #176  

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Release | Other) 

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/app/src/main/java/org/smartregister/fhircore/fragment/PatientDetailFragment.kt [hunk 1]
```diff
@@ -140,6 +142,14 @@ class PatientDetailFragment : Fragment() {
               nextDoseNumber,
               nextVaccineDate
             )
+          val cal: Calendar = Calendar.getInstance()
+          cal.add(Calendar.DATE, -28)
```
[the remark this answers is not in the document]
Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/fragment/PatientDetailFragment.kt]: fixed.

**Comment by syedowaisali90** [comment]
> needs test

we need to merge the `main` branch into this branch because i already added a number of test classes in the `main` branch and require those test classes for wrting more test cases but now the current base branch is `covax-demo`

File: android/app/src/main/java/org/smartregister/fhircore/fragment/PatientDetailFragment.kt [hunk 2]
```diff
@@ -140,6 +142,14 @@ class PatientDetailFragment : Fragment() {
...
+          val overDueStart = cal.time
+          if (immunization.recorded.before(overDueStart)) {
```
Comment (vincent-karuri) [inline android/app/src/main/java/org/smartregister/fhircore/fragment/PatientDetailFragment.kt]: Checking overdue vaccines may become a common operation, could we move it into a util or private method `isOverdue` that checks something like: 

```
immunization.recorded.add(28).before(Calendar.getInstance().time)
```

of course replacing the 28 with a function param.

File: android/app/src/main/java/org/smartregister/fhircore/fragment/PatientListFragment.kt [hunk 3]
```diff
@@ -111,6 +114,15 @@ class PatientListFragment : Fragment() {
       }
     )
 
+    btnOverdue.setOnCheckedChangeListener { buttonView, isChecked ->
+      patientListViewModel.searchResults(
```
Comment (vincent-karuri) [inline android/app/src/main/java/org/smartregister/fhircore/fragment/PatientListFragment.kt:118]: I think this is simply supposed to set a sharedpreference boolean that can be later accessed here https://github.com/opensrp/fhircore/pull/194/files#diff-14402c85a380eedd89160390f84f7c082323e0c97d31388c3d3c43e278d53a94R95.

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt [hunk 4]
```diff
@@ -73,7 +73,12 @@ class PatientListViewModel(application: Application, private val fhirEngine: Fhi
     MutableLiveData<List<Immunization>>()
   }
 
-  fun searchResults(query: String? = null, page: Int = 0, pageSize: Int = 10) {
+  fun searchResults(
+    query: String? = null,
+    page: Int = 0,
+    pageSize: Int = 10,
+    showOverdue: Boolean = true
```
Comment (vincent-karuri) [inline android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt:81]: I don't think we need to modify this fun signature

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt [hunk 5]
```diff
@@ -142,6 +151,19 @@ class PatientListViewModel(application: Application, private val fhirEngine: Fhi
     return (filter == null || filter.equals(this.idElement.idPart))
   }
 
+  protected suspend fun Patient.canAddOverdue(showOverdue: Boolean): Boolean {
```
Comment (vincent-karuri) [inline android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt:154]: what does this function do? not sure I understand

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt (hunk 4 above)
Comment (vincent-karuri) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt:81]: https://github.com/opensrp/fhircore/pull/194/files#r645400196

**Review by vincent-karuri — DISMISSED** [review]
Please address the comments and ensure all items in the checklist are completed

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt (hunk 5 above)
Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt:154]: actually, this function is checking that patient would be shown or not in the patient list like if the `showOverdue` param will `true` then we will show all the patient in the list either the patient second vaccination date is overdue or not and if the `showOverdue` param will `false` then we will hide all those patient who have not received vaccination after the specific time.

File: android/app/src/main/java/org/smartregister/fhircore/fragment/PatientDetailFragment.kt (hunk 2 above)
Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/fragment/PatientDetailFragment.kt]: logic moved in Util.

File: android/app/src/main/java/org/smartregister/fhircore/fragment/PatientListFragment.kt (hunk 3 above)
Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/fragment/PatientListFragment.kt]: fixed. now using SP

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt (hunk 4 above)
Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt]: fixed. showOverdue param removed.

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt (hunk 5 above)
Comment (vincent-karuri) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt]: still not sure I follow, if it's checking status, can't this work https://github.com/opensrp/fhircore/blob/d4dd6042be7c4594aee4213d7a36e165219e90d6/android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt#L113?

Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt]: could you elaborate a bit more on what's not working?

**Comment by pld** [comment]
> > needs test
> 
> we need to merge the `main` branch into this branch because i already added a number of test classes in the `main` branch and require those test classes for wrting more test cases but now the current base branch is `covax-demo`

sounds good! please rebase this over main

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt (hunk 5 above)
Comment (vincent-karuri) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt]: A filter needs to be applied to display only patients with overdue vaccines  when the `Show overdue` button is toggled on. Logic to filter out overdue patients is already implemented https://github.com/opensrp/fhircore/pull/194#discussion_r645452420. Why not use that when filtering here?

Comment (vincent-karuri) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt]: When the button is toggled off, we should display all patients - both overdue and not.

Comment (syedowaisali90) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt]: fixed.

**Review by pld — DISMISSED** [review]

**Comment by vincent-karuri** [comment]
@ndegwamartin please confirm if we should be merging these changes into main

**Review by pld — APPROVED** [review]

File: android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt [hunk 6]
```diff
@@ -60,26 +71,35 @@ class PatientListViewModel(application: Application, private val fhirEngine: Fhi
...
           sort(Patient.GIVEN, Order.ASCENDING)
-          count = pageSize
-          from = (page * pageSize)
         }
+      val showOnlyOverdue = SharedPrefrencesHelper.read(PatientListFragment.SHOW_OVERDUE_PATIENTS)
+      searchResults = searchResults.distinctBy { it.logicalId }
+      searchResults =
+        searchResults.filter {
+          (it.nameMatchesFilter(query) || it.idMatchesFilter(query)) &&
+            if (!showOnlyOverdue) true
```
Comment (vincent-karuri) [inline android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt:94]: This if statement is not necessary

Comment (pld) [reply android/app/src/main/java/org/smartregister/fhircore/viewmodel/PatientListViewModel.kt:94]: agree I'll open a fix PR

**Comment by ndegwamartin** [comment]
@pld I think we need to revert this and tag it with a **DNM** . It appears to have been branched off the `covax-demo` branch and not from `main`. It contains a lot of hardcoded stuff and TODOs that are part of the ongoing clean up this week.

**Pull request merged**